fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java

```diff
@@ -156,26 +165,71 @@ public static FHIRValidator validator() {
+     * @param profileReferences
+     *     the list of profile references to validate
+     * @param resourceAsserted
+     *     indicates whether the profile references came from the resource or were explicitly passed in as arguments
+     * @param issues
+     *     the list of issues to add to
+     */
+    private static void validateProfileReferences(
+            FHIRPathResourceNode resourceNode,
+            List<String> profileReferences,
+            boolean resourceAsserted,
+            List<Issue> issues) {
+        Class<?> resourceType = resourceNode.resource().getClass();
+        for (String url : profileReferences) {
+            StructureDefinition profile = ProfileSupport.getProfile(url);
+            if (profile == null) {
+                issues.add(Issue.builder()
+                    .severity(resourceAsserted ? IssueSeverity.WARNING : IssueSeverity.ERROR)
+                    .code(IssueType.NOT_SUPPORTED)
+                    .details(CodeableConcept.builder()
+                        .text(string("Profile: " + url + " is not supported"))
+                        .build())
+                    .expression(string(resourceNode.path()))
+                    .build());
+            } else if (!ProfileSupport.isApplicable(profile, resourceType)) {
+                issues.add(Issue.builder()
+                    .severity(IssueSeverity.ERROR)
+                    .code(IssueType.INVALID)
+                    .details(CodeableConcept.builder()
+                        .text(string("Profile: " + url + " is not applicable to resource type: " + resourceType.getSimpleName()))
+                        .build())
+                    .expression(string(resourceNode.path()))
+                    .build());
+            }
+        }
+    }
 
-        private ValidatingNodeVisitor() { }
+    private static class ValidatingNodeVisitor extends FHIRPathDefaultNodeVisitor {
+        private final FHIRPathEvaluator evaluator = FHIRPathEvaluator.evaluator();
+        private final EvaluationContext evaluationContext;
+        private final boolean includeResourceAssertedProfiles;
+        private final List<String> profiles;
+        private final List<Issue> issues = new ArrayList<>();
 
-        private List<Issue> validate(EvaluationContext evaluationContext, boolean includeResourceAssertedProfiles, String... profiles) {
-            reset();
+        private ValidatingNodeVisitor(EvaluationContext evaluationContext, boolean includeResourceAssertedProfiles, List<String> profiles) {
             this.evaluationContext = evaluationContext;
             this.includeResourceAssertedProfiles = includeResourceAssertedProfiles;
-            this.profiles = Arrays.asList(profiles);
-            this.evaluationContext.getTree().getRoot().accept(this);
-            return Collections.unmodifiableList(issues);
+            this.profiles = profiles;
         }
 
-        private void reset() {
-            issues.clear();
```

Comment:
why no more reset?

Reply:
We create a visitor each call to FHIRValidator.validate. No need to reset anything.